File: java_refactoring/src/main/java/org/example2/solution/GreenTeaService.java
package org.example2.solution;

public class GreenTeaService extends TeaService {
    public GreenTeaService(int weight) {
        super(weight);
    }

    @Override
    protected TeaType getTeaType() {
        return TeaType.GREEN;
    }

    @Override
    protected void makeTeaType() {
        System.out.println("Having green tea");
        addAmountMade(getWeight() * .16 - .1);
        if (getTeaStrength() > 0) {
            System.out.println("Adding extra strength to green tea");
            addTeaStrength(1.1);
        }
    }

    @Override
    protected void pourTeaType() {
        System.out.println("Pouring green tea");
        System.out.println("Green tea has its origins in ancient China, where it has been consumed for thousands of years, particularly during the Tang and Song dynasties.");
        addAmountMade(-1.2);
    }
}

File: typescript_refactoring/example2/solution/GreenTeaService.ts
import { TeaService } from "./TeaService";
import { TeaType } from "./TeaTypes";

export class GreenTeaService extends TeaService {
    protected readonly teaType = TeaType.Green;

    public constructor(weight: number) {
        super(weight);
    }

    protected makeTeaType(): void {
        console.log("Having green tea");
        this.amountMade += this.weight * 0.16 - 0.1;
        if (this.teaStrength > 0) {
            console.log("Adding extra strength to green tea");
            this.teaStrength += 1.1;
        }
    }

    protected pourTeaType(): void {
        console.log("Pouring green tea");
        console.log("Green tea has its origins in ancient China, where it has been consumed for thousands of years, particularly during the Tang and Song dynasties.");
        this.amountMade -= 1.2;
    }
}
